rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java

```diff
@@ -328,52 +338,148 @@ protected TxType getTransactionType(BtcTransaction btcTx) {
         return TxType.UNKNOWN;
     }
 
-    protected void processPegIn(BtcTransaction btcTx, Transaction rskTx, int height, Sha256Hash btcTxHash)
-            throws IOException, RegisterBtcTransactionException {
+    protected void processPegIn(
+        BtcTransaction btcTx,
+        Transaction rskTx,
+        int height,
+        Sha256Hash btcTxHash) throws IOException, RegisterBtcTransactionException {
+
         logger.debug("[processPegIn] This is a lock tx {}", btcTx);
-        Optional<BtcLockSender> btcLockSenderOptional = btcLockSenderProvider.tryGetBtcLockSender(btcTx);
-        if (!btcLockSenderOptional.isPresent() ||
-                !BridgeUtils.txIsProcessable(btcLockSenderOptional.get().getType(), activations)) {
-            logger.warn("[processPegIn] [btcTx:{}] Could not get BtcLockSender from Btc tx", btcTx.getHash());
-            throw new RegisterBtcTransactionException("Could not get BtcLockSender from Btc tx");
+
+        Coin totalAmount = computeTotalAmountSent(btcTx);
+
+        PeginInformation peginInformation = new PeginInformation(
+            btcLockSenderProvider,
+            peginInstructionsProvider,
+            activations
+        );
+        try {
+            peginInformation.parse(btcTx);
+        } catch (PeginInstructionsException e) {
+            if (activations.isActive(ConsensusRule.RSKIP170)) {
+                // If possible to get the sender address, refund
+                refundTxSender(btcTx, rskTx, peginInformation, totalAmount);
```

Comment:
Here: refund is only created when there's a pegininstructionexception only after the rskip170 activation